File: go/0079_word_search.go
func exist(board [][]byte, word string) bool {
	m, n := len(board), len(board[0])
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			p := board[i][j]
			if p == word[0] {
				res := false
				search(&res, i, j, m, n, board, word, 0)
				if res {
					return true
				}
			}
		}
	}
	return false
}

func search(res *bool, i, j, m, n int, board [][]byte, word string, currentIndex int) {
	var zeroByte byte
	if *res || word[currentIndex] != board[i][j] || board[i][j] == zeroByte {
		return
	}
	if currentIndex == len(word)-1 {
		*res = true
		return
	}

	nextIndex := currentIndex + 1
	tmp := board[i][j]
	board[i][j] = zeroByte
	if i > 0 {
		search(res, i-1, j, m, n, board, word, nextIndex)
	}
	if i < m-1 {
		search(res, i+1, j, m, n, board, word, nextIndex)
	}
	if j > 0 {
		search(res, i, j-1, m, n, board, word, nextIndex)
	}
	if j < n-1 {
		search(res, i, j+1, m, n, board, word, nextIndex)
	}
	board[i][j] = tmp
}

File: python/0079_word_search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        if not word:
            return False

        mlen = len(board)
        if mlen == 0:
            return False
        nlen = len(board[0])
        if nlen == 0:
            return False

        visited = set()
        for i in range(mlen):
            for j in range(nlen):
                visited.add((i, j))
                if board[i][j] == word[0] and search(board, i, j, word, 1, visited):
                    return True
                visited.remove((i, j))
        return False


def search(board, i, j, word, index, visited):
    if index == len(word):
        return True

    for adj in [(i - 1, j), (i + 1, j), (i, j - 1), (i, j + 1)]:
        if adj[0] >= 0 and adj[0] < len(board) and adj[1] >= 0 and adj[1] < len(board[0]):
            if board[adj[0]][adj[1]] == word[index] and adj not in visited:
                visited.add(adj)
                if search(board, adj[0], adj[1], word, index + 1, visited):
                    return True
                visited.remove(adj)
